File: Python3/1001 - 1200/1195 Fizz Buzz Multithreaded.py
# 32ms, 87.09%; 13.1MB, 100.00%
class FizzBuzz:
    def __init__(self, n: int):
        self.n = n + 1
        self.f = threading.Semaphore(0)
        self.b = threading.Semaphore(0)
        self.fb = threading.Semaphore(0)
        self.num = threading.Semaphore(1)
        self.cur = 1

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: callable) -> None:
        for i in range(1, self.n):
            if i % 5 and i % 3 == 0:
                self.f.acquire()
                printFizz()
                self.cur += 1
                if self.cur % 5:
                    self.num.release()
                else:
                    self.b.release()

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: callable) -> None:
        for i in range(1, self.n):
            if i % 5 == 0 and i % 3:
                self.b.acquire()
                printBuzz()
                self.cur += 1
                if self.cur % 3:
                    self.num.release()
                else:
                    self.f.release()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: callable) -> None:
        for i in range(1, self.n):
            if i % 5 == 0 and i % 3 == 0:
                self.fb.acquire()
                printFizzBuzz()
                self.cur += 1
                self.num.release()

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: callable) -> None:
        for i in range(1, self.n):
            if i % 5 and i % 3:
                self.num.acquire()
                printNumber(i)
                self.cur += 1
                if self.cur % 5 == 0 and self.cur % 3 == 0:
                    self.fb.release()
                elif self.cur % 3 == 0:
                    self.f.release()
                elif self.cur % 5 == 0:
                    self.b.release()
                else:
                    self.num.release()

File: 1001 - 1200/1195. Fizz Buzz Multithreaded.cc
// 48ms, 57.48%; 9.1MB, 100.00%
#include <bits/stdc++.h>
using namespace std;

class FizzBuzz {
private:
    int n;
    int cur;
    mutex mx;
    condition_variable cv;

public:
    FizzBuzz(int n) {
        this->n = n;
        this->cur = 1;
    }

    // printFizz() outputs "fizz".
    void fizz(function<void()> printFizz) {
        helper([&]() { return cur % 5 && cur % 3 == 0; }, [&](int i) { printFizz(); });
    }

    // printBuzz() outputs "buzz".
    void buzz(function<void()> printBuzz) {
        helper([&]() { return cur % 5 == 0 && cur % 3; }, [&](int i) { printBuzz(); });
    }

    // printFizzBuzz() outputs "fizzbuzz".
    void fizzbuzz(function<void()> printFizzBuzz) {
        helper([&]() { return cur % 5 == 0 && cur % 3 == 0; }, [&](int i) { printFizzBuzz(); });
    }

    // printNumber(x) outputs "x", where x is an integer.
    void number(function<void(int)> printNumber) {
        helper([&]() { return cur % 5 && cur % 3; }, printNumber);
    }

private:
    void helper(function<bool()> test, function<void(int)> print) {
        while (true) {
            unique_lock<mutex> lock(mx);
            cv.wait(lock, [&]() { return cur > n || test(); });
            if (cur > n) break;
            print(cur++);
            lock.unlock();
            cv.notify_all();
        }
    }
};
